File: Oving1/go/foo.go
// Use `go run foo.go` to run your program

package main

import (
	. "fmt"
	"runtime"
	"time"
)

var i = 0

func incrementing() {
	//TODO: increment i 1000000 times
	for j := 0; j < 1000000; j++ {
		i++
	}
}

func decrementing() {
	//TODO: decrement i 1000000 times
	for k := 0; k < 1000000; k++ {
		i--
	}
}

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU()) // I guess this is a hint to what GOMAXPROCS does...
	// Try doing the exercise both with and without it!

	// TODO: Spawn both functions as goroutines
	go incrementing()
	go decrementing()

	// We have no way to wait for the completion of a goroutine (without additional syncronization of some sort)
	// We'll come back to using channels in Exercise 2. For now: Sleep.
	time.Sleep(500 * time.Millisecond)
	Println("The magic number is:", i)
}

//What is happening? The two functions is running parallell. That means there is a chance that incrementing will read i=1, 
//then decrementing read i = 1, since incrementing has not written i=2 back. Then incrementing will continoue i++ and then write i = 2 back,
// but at the same time decrementing is computing with i = 1 and i-- to i = 0 and write back i = 0; We have now lost the incrementation in addition to running decrementing one time.
// This can happen several times and that is why we get seamingly random numbers of positive or negative sign.

File: Oving1/c/foo.c
// Compile with `gcc foo.c -std=c99 -lpthread`, or use the makefile

#include <pthread.h>
#include <stdio.h>

int i = 0;

// Note the return type: void*
void* incrementingThreadFunction(){
    // TODO: increment i 1_000_000 times
    for(int j = 1; j < 1000000; j++){
        i++;
    }
    return NULL;
}

void* decrementingThreadFunction(){
    // TODO: decrement i 1_000_000 times
    for(int k = 1; k < 1000000; k++){
        i--;
    }
    return NULL;
}


int main(){
    // TODO: declare incrementingThread and decrementingThread (hint: google pthread_create)
    pthread_t incrementingThread;
    pthread_t decrementingThread;

    
    pthread_create(&incrementingThread, NULL, incrementingThreadFunction, NULL);
    pthread_create(&decrementingThread, NULL, decrementingThreadFunction, NULL);
    
    pthread_join(incrementingThread, NULL);
    pthread_join(decrementingThread, NULL);
    
    printf("The magic number is: %d\n", i);
    return 0;
}
